File: js/ai/src/session-stores.ts
/**
 * Copyright 2026 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { GenkitError } from '@genkit-ai/core';
import * as fs from 'fs';
import * as fsp from 'fs/promises';
import * as path from 'path';
import {
  assertValidSessionId,
  type GetSnapshotOptions,
  type SessionSnapshot,
  type SessionStore,
  type SessionStoreOptions,
  type SnapshotMutator,
} from './session.js';

/**
 * Normalizes and validates {@link GetSnapshotOptions}.
 *
 * Enforces that exactly one of `snapshotId` / `sessionId` is provided and,
 * when a `sessionId` is given, that it is a valid UUID.
 *
 * @throws `INVALID_ARGUMENT` when neither or both are provided.
 */
function normalizeGetSnapshotOptions(opts: GetSnapshotOptions): {
  snapshotId?: string;
  sessionId?: string;
} {
  const { snapshotId, sessionId } = opts;
  if (!!snapshotId === !!sessionId) {
    throw new GenkitError({
      status: 'INVALID_ARGUMENT',
      message:
        `getSnapshot requires exactly one of 'snapshotId' or 'sessionId' ` +
        `(got ${snapshotId ? 'snapshotId' : 'neither'}${sessionId ? ' and sessionId' : ''}).`,
    });
  }
  if (sessionId) {
    assertValidSessionId(sessionId);
  }
  return { snapshotId, sessionId };
}

/**
 * Selects the latest leaf snapshot from a set belonging to one session.
 *
 * A "leaf" is a snapshot that no other snapshot points to as its `parentId`.
 * A healthy linear session has exactly one leaf - the latest turn.
 *
 * - Returns `undefined` when `snapshots` is empty.
 * - Returns the single leaf when the history is linear.
 * - When the history has branched (more than one leaf, e.g. after a
 *   regenerate) the behavior depends on `rejectBranching`:
 *   - `false` (default): returns the most-recently created leaf (by
 *     `createdAt`). This keeps `sessionId` lookups cheap and forgiving.
 *   - `true`: throws `FAILED_PRECONDITION`, since there is no unambiguous
 *     "latest". Opt into this in dev to surface accidental branching early.
 */
function selectLeafSnapshot<S>(
  snapshots: SessionSnapshot<S>[],
  sessionId: string,
  rejectBranching = false
): SessionSnapshot<S> | undefined {
  if (snapshots.length === 0) return undefined;

  const parentIds = new Set<string>();
  for (const snap of snapshots) {
    if (snap.parentId) parentIds.add(snap.parentId);
  }
  const leaves = snapshots.filter((s) => !parentIds.has(s.snapshotId));

  // A single-snapshot session, or any chain, collapses to one leaf.
  if (leaves.length === 1) return leaves[0];

  if (leaves.length === 0) {
    // Cyclic / corrupt history - every snapshot is someone's parent.
    throw new GenkitError({
      status: 'FAILED_PRECONDITION',
      message:
        `Session '${sessionId}' has no leaf snapshot (corrupt or cyclic ` +
        `history). Resume by snapshotId instead.`,
    });
  }

  if (rejectBranching) {
    throw new GenkitError({
      status: 'FAILED_PRECONDITION',
      message:
        `Session '${sessionId}' has branching snapshots (${leaves.length} ` +
        `leaves), so there is no single latest snapshot. This happens when a ` +
        `conversation is branched (e.g. regenerate). Resume by snapshotId instead.`,
    });
  }

  // Default: pick the most-recently created leaf. `createdAt` is an ISO-8601
  // timestamp, so lexicographic comparison matches chronological order.
  return leaves.reduce((latest, snap) =>
    snap.createdAt > latest.createdAt ? snap : latest
  );
}

/**
 * In-memory implementation of persistent Session Store.
 */
export class InMemorySessionStore<S = unknown> implements SessionStore<S> {
  private snapshots = new Map<string, SessionSnapshot<S>>();
  private listeners = new Map<
    string,
    Array<(snapshot: SessionSnapshot<S>) => void>
  >();
  private rejectBranchingSessions: boolean;

  /**
   * @param options.rejectBranchingSessions When `true`, a `sessionId` lookup
   *   that resolves to a branched history (more than one leaf) throws
   *   `FAILED_PRECONDITION` instead of returning the latest leaf. Defaults to
   *   `false`; opt in (e.g. in dev) to surface accidental branching early.
   */
  constructor(options?: { rejectBranchingSessions?: boolean }) {
    this.rejectBranchingSessions = options?.rejectBranchingSessions ?? false;
  }

  async getSnapshot(
    opts: GetSnapshotOptions
  ): Promise<SessionSnapshot<S> | undefined> {
    const { snapshotId, sessionId } = normalizeGetSnapshotOptions(opts);

    if (snapshotId) {
      const snap = this.snapshots.get(snapshotId);
      if (!snap) return undefined;
      return structuredClone(snap);
    }

    // sessionId lookup: gather every snapshot belonging to this session and
    // resolve the single leaf (latest) snapshot.
    const owned: SessionSnapshot<S>[] = [];
    for (const snap of this.snapshots.values()) {
      if (snap.state?.sessionId === sessionId) {
        owned.push(snap);
      }
    }
    const leaf = selectLeafSnapshot(
      owned,
      sessionId!,
      this.rejectBranchingSessions
    );
    return leaf ? structuredClone(leaf) : undefined;
  }

  async saveSnapshot(
    snapshotId: string | undefined,
    mutator: SnapshotMutator<S>,
    options?: SessionStoreOptions
  ): Promise<string | null> {
    const current = snapshotId ? this.snapshots.get(snapshotId) : undefined;

    const result = mutator(current ? structuredClone(current) : undefined);
    if (result === null) return null;

    // Determine the final ID. The runtime normally supplies a snapshotId, but
    // fall back to a fresh UUID for direct store users who omit it.
    const id =
      snapshotId || result.snapshotId || globalThis.crypto.randomUUID();
    const full: SessionSnapshot<S> = {
      ...result,
      snapshotId: id,
    };

    this.snapshots.set(id, structuredClone(full));
    const snapshotListeners = this.listeners.get(id);
    if (snapshotListeners) {
      for (const listener of snapshotListeners) {
        listener(structuredClone(full));
      }
    }
    return id;
  }

  onSnapshotStateChange(
    snapshotId: string,
    callback: (snapshot: SessionSnapshot<S>) => void,
    options?: SessionStoreOptions
  ): void | (() => void) {
    if (!this.listeners.has(snapshotId)) {
      this.listeners.set(snapshotId, []);
    }
    this.listeners.get(snapshotId)!.push(callback);
    return () => {
      const list = this.listeners.get(snapshotId);
      if (list) {
        const index = list.indexOf(callback);
        if (index >= 0) list.splice(index, 1);
      }
    };
  }
}

/**
 * A Node.js file-system backed session snapshot store.
 *
 * Snapshots are stored as flat JSON files keyed by their `snapshotId`, under an
 * optional per-tenant sub-directory `prefix`:
 *
 * File layout: `dirPath/<prefix>/<snapshotId>.json`
 *
 * `getSnapshot({ sessionId })` scans the prefix directory and selects the
 * single leaf snapshot whose `state.sessionId` matches - there is no separate
 * grouping directory, the `sessionId` lives in each snapshot's state.
 */
export class FileSessionStore<S = unknown> implements SessionStore<S> {
  private dirPath: string;
  private maxPersistedChainLength?: number;
  private snapshotPathPrefix?: (options?: SessionStoreOptions) => string;
  private rejectBranchingSessions: boolean;

  /**
   * @param dirPath Directory where snapshot JSON files are stored.
   * @param options.maxPersistedChainLength When set, snapshots older than this
   *   many entries in a chain are automatically deleted on each save.
   * @param options.snapshotPathPrefix Returns a sub-directory prefix derived
   *   from the call's {@link SessionStoreOptions} (e.g. the authenticated user
   *   id from `options.context`), useful for multi-tenant isolation: all reads
   *   and writes are scoped to that prefix, so one tenant can never see
   *   another's snapshots. Defaults to `"global"`.
   * @param options.rejectBranchingSessions When `true`, a `sessionId` lookup
   *   that resolves to a branched history (more than one leaf) throws
   *   `FAILED_PRECONDITION` instead of returning the latest leaf. Defaults to
   *   `false`; opt in (e.g. in dev) to surface accidental branching early.
   */
  constructor(
    dirPath: string,
    options?: {
      maxPersistedChainLength?: number;
      snapshotPathPrefix?: (options?: SessionStoreOptions) => string;
      rejectBranchingSessions?: boolean;
    }
  ) {
    this.dirPath = path.resolve(dirPath);
    fs.mkdirSync(this.dirPath, { recursive: true });
    this.maxPersistedChainLength = options?.maxPersistedChainLength;
    this.snapshotPathPrefix = options?.snapshotPathPrefix;
    this.rejectBranchingSessions = options?.rejectBranchingSessions ?? false;
  }

  private async ensureDir(dir: string): Promise<void> {
    await fsp.mkdir(dir, { recursive: true });
  }

  /** Resolves the (per-tenant) directory snapshots are stored under. */
  private prefixDir(options?: SessionStoreOptions): string {
    const prefix = this.snapshotPathPrefix
      ? this.snapshotPathPrefix(options)
      : 'global';
    return path.join(this.dirPath, prefix);
  }

  /**
   * Resolves the file path for a given snapshotId: `<prefix>/<snapshotId>.json`.
   */
  private async getFilePath(
    snapshotId: string,
    options?: SessionStoreOptions
  ): Promise<string> {
    const dir = this.prefixDir(options);
    await this.ensureDir(dir);
    return path.join(dir, `${snapshotId}.json`);
  }

  async getSnapshot(
    opts: GetSnapshotOptions
  ): Promise<SessionSnapshot<S> | undefined> {
    const { snapshotId, sessionId } = normalizeGetSnapshotOptions(opts);

    if (sessionId) {
      return this.getLatestSnapshotForSession(sessionId, opts);
    }

    const filePath = await this.getFilePath(snapshotId!, opts);
    try {
      const fileContents = await fsp.readFile(filePath, 'utf-8');
      return JSON.parse(fileContents) as SessionSnapshot<S>;
    } catch (e: unknown) {
      if ((e as NodeJS.ErrnoException).code === 'ENOENT') return undefined;
      throw e;
    }
  }

  /**
   * Loads a single snapshot file by its id (no sessionId branch). Used by
   * internal traversal (parent chains) where we always have a concrete id.
   */
  private async getSnapshotById(
    snapshotId: string,
    options?: SessionStoreOptions
  ): Promise<SessionSnapshot<S> | undefined> {
    const filePath = await this.getFilePath(snapshotId, options);
    try {
      const fileContents = await fsp.readFile(filePath, 'utf-8');
      return JSON.parse(fileContents) as SessionSnapshot<S>;
    } catch (e: unknown) {
      if ((e as NodeJS.ErrnoException).code === 'ENOENT') return undefined;
      throw e;
    }
  }

  /**
   * Resolves the latest (leaf) snapshot for a session by scanning every
   * snapshot file in the (per-tenant) prefix directory, keeping those whose
   * `state.sessionId` matches, and selecting the single leaf.
   */
  private async getLatestSnapshotForSession(
    sessionId: string,
    options?: SessionStoreOptions
  ): Promise<SessionSnapshot<S> | undefined> {
    const dir = this.prefixDir(options);

    let files: string[];
    try {
      files = await fsp.readdir(dir);
    } catch (e: unknown) {
      if ((e as NodeJS.ErrnoException).code === 'ENOENT') return undefined;
      throw e;
    }

    const snapshots: SessionSnapshot<S>[] = [];
    for (const file of files) {
      if (!file.endsWith('.json')) continue;
      try {
        const contents = await fsp.readFile(path.join(dir, file), 'utf-8');
        const snap = JSON.parse(contents) as SessionSnapshot<S>;
        if (snap.state?.sessionId === sessionId) {
          snapshots.push(snap);
        }
      } catch (e: unknown) {
        if ((e as NodeJS.ErrnoException).code === 'ENOENT') continue;
        throw e;
      }
    }

    return selectLeafSnapshot(
      snapshots,
      sessionId,
      this.rejectBranchingSessions
    );
  }

  async saveSnapshot(
    snapshotId: string | undefined,
    mutator: SnapshotMutator<S>,
    options?: SessionStoreOptions
  ): Promise<string | null> {
    // Read the current snapshot when an ID is provided.
    const current = snapshotId
      ? await this.getSnapshotById(snapshotId, options)
      : undefined;

    const snapshot = mutator(current);
    if (snapshot === null) return null;

    // Determine the final ID. The runtime normally supplies a snapshotId, but
    // fall back to a fresh UUID for direct store users who omit it.
    const id =
      snapshotId || snapshot.snapshotId || globalThis.crypto.randomUUID();

    const full: SessionSnapshot<S> = {
      ...snapshot,
      snapshotId: id,
    };
    const filePath = await this.getFilePath(id, options);
    await fsp.writeFile(filePath, JSON.stringify(full, null, 2), 'utf-8');

    if (this.maxPersistedChainLength && this.maxPersistedChainLength > 0) {
      let cur: SessionSnapshot<S> | undefined = full;
      const chain: string[] = [];

      while (cur) {
        chain.push(cur.snapshotId);
        if (cur.parentId) {
          cur = await this.getSnapshotById(cur.parentId, options);
        } else {
          break;
        }
      }

      if (chain.length > this.maxPersistedChainLength) {
        for (let i = this.maxPersistedChainLength; i < chain.length; i++) {
          const pathToDelete = await this.getFilePath(chain[i], options);
          await fsp.unlink(pathToDelete).catch((e: unknown) => {
            if ((e as NodeJS.ErrnoException).code !== 'ENOENT') throw e;
          });
        }
      }
    }

    return id;
  }
}

File: py/packages/genkit/src/genkit/_ai/_agents/_session_stores.py
# Copyright 2026 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Flat, full-snapshot session stores for local development and tests.

Each turn is persisted whole, keyed by its snapshot id, and the ``parent_id``
links between snapshots form the conversation tree. That single shape covers a
linear chat, a "just give me the latest turn" lookup, and a forked/branching
history without any diffing — so it's the right default for local dev, tests,
and single-process apps. For a multi-instance production deployment, back the
same ``SessionStore`` protocol with a real database (where it's worth trading
this simplicity for incremental, diff-based persistence).

``InMemorySessionStore`` and ``FileSessionStore`` are standalone — they share
the storage-agnostic bits below as plain functions rather than a common base
class, so each store is a self-contained read of how its backend works.
"""

from __future__ import annotations

import asyncio
import os
from collections.abc import Callable
from datetime import datetime, timezone
from typing import Generic
from uuid import uuid4

from genkit._ai._agents._session import (
    SessionStore,
    SnapshotSubscriber,
    StateT,
    select_leaf_snapshot,
)
from genkit._core._error import GenkitError
from genkit._core._typing import SessionSnapshot, SnapshotStatus

SaveFn = Callable[[SessionSnapshot | None], SessionSnapshot | None]
Subs = dict[str, list['asyncio.Queue[SnapshotStatus | None]']]


def session_id_of(snapshot: SessionSnapshot) -> str | None:
    """Session a snapshot belongs to, preferring the top-level id over state's."""
    if snapshot.session_id:
        return snapshot.session_id
    return snapshot.state.session_id if snapshot.state is not None else None


def require_one_selector(*, snapshot_id: str | None, session_id: str | None) -> None:
    """Enforce that a get_snapshot call names exactly one of snapshot_id / session_id."""
    if bool(snapshot_id) == bool(session_id):
        raise GenkitError(
            status='INVALID_ARGUMENT',
            message=(
                "get_snapshot requires exactly one of 'snapshot_id' or "
                f"'session_id' (got {'snapshot_id' if snapshot_id else 'neither'}"
                f'{" and session_id" if session_id else ""}).'
            ),
        )


def select_leaf(
    *,
    snapshots: list[SessionSnapshot],
    session_id: str,
    reject_ambiguous: bool,
) -> SessionSnapshot | None:
    """Resolve a session's current leaf from all its snapshots.

    A leaf is a snapshot no other snapshot names as a parent. A linear chat has
    exactly one; a forked history has several. When opted in we reject the
    ambiguous case, otherwise the most recently created leaf wins so a sibling
    left behind by an aborted/failed turn never shadows the live one.
    """
    if not snapshots:
        return None

    if reject_ambiguous:
        return select_leaf_snapshot(snapshots=snapshots, session_id=session_id)

    parent_ids = {snap.parent_id for snap in snapshots if snap.parent_id}
    leaves = [snap for snap in snapshots if snap.snapshot_id not in parent_ids]
    if not leaves:
        raise GenkitError(
            status='FAILED_PRECONDITION',
            message=(
                f"Session '{session_id}' has no leaf snapshot (corrupt or cyclic "
                'history). Resume by snapshot_id instead.'
            ),
        )
    # created_at is an ISO-8601 string, so lexicographic max is chronological;
    # snapshot_id breaks exact ties deterministically.
    return max(leaves, key=lambda snap: (snap.created_at, snap.snapshot_id))


def stamp_store_fields(*, snapshot: SessionSnapshot, snapshot_id: str | None) -> None:
    """Fill in the fields the store owns on a snapshot about to be written."""
    snapshot.snapshot_id = snapshot_id or str(uuid4())
    if not snapshot.created_at:
        snapshot.created_at = datetime.now(timezone.utc).isoformat()
    if not snapshot.status:
        snapshot.status = SnapshotStatus.COMPLETED
    # Mirror the session id up to the top level so session lookups and callers
    # reading snapshot.session_id don't have to dig into state.
    if not snapshot.session_id and snapshot.state is not None:
        snapshot.session_id = snapshot.state.session_id


def apply_save(*, existing: SessionSnapshot | None, snapshot_id: str | None, fn: SaveFn) -> SessionSnapshot | None:
    """Run a save mutator and stamp the result, or None to skip the write."""
    if snapshot_id is not None and existing is None:
        return None
    next_snapshot = fn(existing.model_copy(deep=True) if existing is not None else None)
    if next_snapshot is None:
        return None
    stamp_store_fields(snapshot=next_snapshot, snapshot_id=snapshot_id)
    return next_snapshot


def notify(*, subs: Subs, snapshot_id: str, status: SnapshotStatus | None) -> None:
    """Push a status change to everyone subscribed to a snapshot."""
    # Subscriber queues are unbounded, so put_nowait can't fail here.
    for q in subs.get(snapshot_id, []):
        q.put_nowait(status)


async def subscribe(
    *,
    subs: Subs,
    snapshot_id: str,
    current: SessionSnapshot | None,
) -> asyncio.Queue[SnapshotStatus | None]:
    """Register a status-change queue, seeding it with the current status."""
    q: asyncio.Queue[SnapshotStatus | None] = asyncio.Queue()
    if current is None:
        await q.put(None)
        return q
    await q.put(current.status)
    subs.setdefault(snapshot_id, []).append(q)
    return q


class InMemorySessionStore(SessionStore[StateT], SnapshotSubscriber, Generic[StateT]):
    """In-memory snapshot store. State is lost when the process exits."""

    def __init__(self, *, reject_ambiguous_session: bool = False) -> None:
        """Create the store.

        When ``reject_ambiguous_session`` is set, a ``session_id`` lookup on a
        history that has forked (more than one leaf) raises instead of picking
        the most recent branch — useful when accidental branching should surface
        loudly.
        """
        self.reject_ambiguous = reject_ambiguous_session
        self.snapshots: dict[str, SessionSnapshot] = {}
        self.subs: Subs = {}

    async def get_snapshot(
        self,
        *,
        snapshot_id: str | None = None,
        session_id: str | None = None,
    ) -> SessionSnapshot | None:
        require_one_selector(snapshot_id=snapshot_id, session_id=session_id)
        async with self.lock:
            if snapshot_id is not None:
                snap = self.snapshots.get(snapshot_id)
                return snap.model_copy(deep=True) if snap is not None else None

            assert session_id is not None
            owned = [snap for snap in self.snapshots.values() if session_id_of(snap) == session_id]
            leaf = select_leaf(snapshots=owned, session_id=session_id, reject_ambiguous=self.reject_ambiguous)
            return leaf.model_copy(deep=True) if leaf is not None else None

    async def save_snapshot(self, snapshot_id: str | None, fn: SaveFn) -> SessionSnapshot | None:
        async with self.lock:
            existing = self.snapshots.get(snapshot_id) if snapshot_id is not None else None
            next_snapshot = apply_save(existing=existing, snapshot_id=snapshot_id, fn=fn)
            if next_snapshot is None:
                return None
            self.snapshots[next_snapshot.snapshot_id] = next_snapshot.model_copy(deep=True)
            notify(subs=self.subs, snapshot_id=next_snapshot.snapshot_id, status=next_snapshot.status)
            return next_snapshot

    async def on_snapshot_status_change(self, snapshot_id: str) -> asyncio.Queue[SnapshotStatus | None]:
        async with self.lock:
            return await subscribe(subs=self.subs, snapshot_id=snapshot_id, current=self.snapshots.get(snapshot_id))


class FileSessionStore(SessionStore[StateT], SnapshotSubscriber, Generic[StateT]):
    """File-backed snapshot store: one ``<snapshot_id>.json`` per snapshot."""

    def __init__(
        self,
        directory: str,
        *,
        reject_ambiguous_session: bool = False,
        max_persisted_chain_length: int | None = None,
    ) -> None:
        """Create the store, ensuring ``directory`` exists.

        See :class:`InMemorySessionStore` for ``reject_ambiguous_session``.

        ``max_persisted_chain_length`` caps how many snapshots of a chat's
        history stay on disk: once a chain grows past it, the oldest turns are
        deleted on each save so a long-lived conversation doesn't accumulate
        files forever. Resuming and continuing still work, but you lose the
        ability to rewind or branch past the retained window. Leave it unset to
        keep the full history.
        """
        self.reject_ambiguous = reject_ambiguous_session
        self.max_persisted_chain_length = max_persisted_chain_length
        self.directory = directory
        self.subs: Subs = {}
        os.makedirs(directory, exist_ok=True)

    def path(self, snapshot_id: str) -> str:
        """Return the file path for a snapshot ID."""
        return os.path.join(self.directory, f'{snapshot_id}.json')

    def read_sync(self, snapshot_id: str) -> SessionSnapshot | None:
        """Read and validate a snapshot from disk synchronously."""
        path = self.path(snapshot_id)
        if not os.path.exists(path):
            return None
        with open(path, encoding='utf-8') as f:
            return SessionSnapshot.model_validate_json(f.read())

    def write_sync(self, snapshot: SessionSnapshot) -> None:
        """Atomically write a snapshot to disk synchronously."""
        path = self.path(snapshot.snapshot_id)
        temp_path = path + '.tmp'
        with open(temp_path, 'w', encoding='utf-8') as f:
            f.write(snapshot.model_dump_json(indent=2))
        os.replace(temp_path, path)

    def delete_sync(self, snapshot_id: str) -> None:
        """Delete a snapshot file, tolerating one that's already gone."""
        try:
            os.remove(self.path(snapshot_id))
        except FileNotFoundError:
            pass

    def prune_chain_sync(self, leaf: SessionSnapshot) -> None:
        """Trim a chat's ancestry to the newest ``max_persisted_chain_length`` turns.

        Walks ``parent_id`` back from the just-written snapshot and deletes the
        oldest links past the cap. The retained oldest snapshot keeps pointing at
        its (now-deleted) parent, so history reconstruction simply stops at the
        window's edge.
        """
        cap = self.max_persisted_chain_length
        if not cap or cap <= 0:
            return
        chain: list[str] = []
        seen: set[str] = set()
        cur: SessionSnapshot | None = leaf
        # `seen` stops a corrupt/cyclic parent chain from looping forever (each
        # hop is a disk read), the same guard walk_back_to_resumable uses.
        while cur is not None and cur.snapshot_id not in seen:
            seen.add(cur.snapshot_id)
            chain.append(cur.snapshot_id)
            cur = self.read_sync(cur.parent_id) if cur.parent_id else None
        for snapshot_id in chain[cap:]:
            self.delete_sync(snapshot_id)

    def read_session_sync(self, session_id: str) -> list[SessionSnapshot]:
        """Read all snapshots for a session from disk synchronously."""
        out: list[SessionSnapshot] = []
        if not os.path.isdir(self.directory):
            return out
        for name in os.listdir(self.directory):
            if not name.endswith('.json'):
                continue
            try:
                with open(os.path.join(self.directory, name), encoding='utf-8') as f:
                    snap = SessionSnapshot.model_validate_json(f.read())
            except (OSError, ValueError):
                continue
            if session_id_of(snap) == session_id:
                out.append(snap)
        return out

    async def get_snapshot(
        self,
        *,
        snapshot_id: str | None = None,
        session_id: str | None = None,
    ) -> SessionSnapshot | None:
        require_one_selector(snapshot_id=snapshot_id, session_id=session_id)
        async with self.lock:
            if snapshot_id is not None:
                return await asyncio.to_thread(self.read_sync, snapshot_id)

            assert session_id is not None
            owned = await asyncio.to_thread(self.read_session_sync, session_id)
            return select_leaf(snapshots=owned, session_id=session_id, reject_ambiguous=self.reject_ambiguous)

    async def save_snapshot(self, snapshot_id: str | None, fn: SaveFn) -> SessionSnapshot | None:
        async with self.lock:
            existing = await asyncio.to_thread(self.read_sync, snapshot_id) if snapshot_id is not None else None
            next_snapshot = apply_save(existing=existing, snapshot_id=snapshot_id, fn=fn)
            if next_snapshot is None:
                return None
            await asyncio.to_thread(self.write_sync, next_snapshot)
            if self.max_persisted_chain_length:
                await asyncio.to_thread(self.prune_chain_sync, next_snapshot)
            notify(subs=self.subs, snapshot_id=next_snapshot.snapshot_id, status=next_snapshot.status)
            return next_snapshot

    async def on_snapshot_status_change(self, snapshot_id: str) -> asyncio.Queue[SnapshotStatus | None]:
        async with self.lock:
            current = await asyncio.to_thread(self.read_sync, snapshot_id)
            return await subscribe(subs=self.subs, snapshot_id=snapshot_id, current=current)
